examples/cis2-wccd-upgradable/src/lib.rs

```diff
@@ -0,0 +1,3181 @@
+                balance:   0u64.into(),
+                operators: state_builder.new_set(),
+            });
+            owner_state.operators.insert(params.operator);
+        }
+        Update::Remove => {
+            host.state_mut().token.entry(params.owner).and_modify(|address_state| {
+                address_state.operators.remove(&params.operator);
+            });
+        }
+    };
+
+    Ok(())
+}
+
+/// Get unpause time.
+#[receive(
+    contract = "CIS2-wCCD-State",
+    name = "getUnpauseTime",
+    return_value = "ReturnUnpauseTime"
+)]
+fn contract_state_get_unpause_time<S: HasStateApi>(
+    _ctx: &impl HasReceiveContext,
+    host: &impl HasHost<State<S>, StateApiType = S>,
+) -> ContractResult<Timestamp> {
+    Ok(host.state().unpause_time)
+}
+
+/// Get Balance.
+#[receive(
+    contract = "CIS2-wCCD-State",
+    name = "getBalance",
+    parameter = "GetBalanceParams",
+    return_value = "ContractTokenAmount"
+)]
+fn contract_state_get_balance<S: HasStateApi>(
+    ctx: &impl HasReceiveContext,
+    host: &impl HasHost<State<S>, StateApiType = S>,
+) -> ContractResult<ContractTokenAmount> {
+    let params: GetBalanceParams = ctx.parameter_cursor().get()?;
+
+    Ok(host.state().token.get(&params.owner).map(|s| s.balance).unwrap_or_else(|| 0u64.into()))
+}
+
+/// Is_operator.
+#[receive(
+    contract = "CIS2-wCCD-State",
+    name = "isOperator",
+    parameter = "IsOperatorParams",
+    return_value = "ReturnIsOperator"
+)]
+fn contract_state_get_operator<S: HasStateApi>(
+    ctx: &impl HasReceiveContext,
+    host: &impl HasHost<State<S>, StateApiType = S>,
+) -> ContractResult<bool> {
+    let params: IsOperatorParams = ctx.parameter_cursor().get()?;
+
+    Ok(host
+        .state()
+        .token
+        .get(&params.owner)
+        .map_or(false, |address_state| address_state.operators.contains(&params.address)))
+}
+
+//// Set current sender.
+#[receive(
+    contract = "CIS2-wCCD",
+    name = "setCurrentSender",
+    parameter = "SetCurrentSenderParams",
+    mutable
+)]
+fn contract_set_current_sender<S: HasStateApi>(
+    ctx: &impl HasReceiveContext,
+    host: &mut impl HasHost<StateImplementation, StateApiType = S>,
+) -> ContractResult<()> {
+    // Only proxy can set this value on the implementation.
+    only_proxy(host.state().proxy_address, ctx.sender())?;
+
+    // Set current sender.
+    let params: SetCurrentSenderParams = ctx.parameter_cursor().get()?;
+    host.state_mut().current_sender = params.sender;
+    Ok(())
+}
+
+/// Function to receive CCD on the implementation.
+#[receive(contract = "CIS2-wCCD", name = "receiveCCD", payable)]
+fn contract_implementation_recieve_ccd<S: HasStateApi>(
+    _ctx: &impl HasReceiveContext,
+    _host: &impl HasHost<StateImplementation, StateApiType = S>,
+    _amount: Amount,
+) -> ContractResult<()> {
+    Ok(())
+}
+
+/// Function to receive CCD on the proxy.
+#[receive(contract = "CIS2-wCCD-Proxy", name = "receiveCCD", payable)]
```

Comment:
Yes, all CCDs are held by the proxy contract. Nonetheless, during the `unwrap` function, we need to get access to some CCD amount from the proxy. Currently, we request in the `unwrap` function that the proxy should send some CCD amount to the implementation which then goes through more complex logic to deal with sending the CCD to the receiver address (which can be an account or a smart contract address) as well as logging required events.

The proxy logic is kept as simple as possible because we cannot upgrade the logic. That is why it only sends CCD to the implementation and then let the implementation code more complex logic.

Reply:
added a comment